src/electrum_fmt_plug.c

```diff
@@ -0,0 +1,314 @@
+		return 0;
+
+	ctcopy = strdup(ciphertext);
+	keeptr = ctcopy;
+
+	ctcopy += TAG_LENGTH;
+	if ((p = strtokm(ctcopy, "*")) == NULL) // type
+		goto err;
+	if (!isdec(p))
+		goto err;
+	value = atoi(p);
+	if (value != 1 && value != 2 && value != 3)
+		goto err;
+	if ((p = strtokm(NULL, "*")) == NULL) // iv
+		goto err;
+	if (hexlenl(p, &extra) != 16 * 2 || extra)
+		goto err;
+	if ((p = strtokm(NULL, "*")) == NULL) // encrypted data (seed part)
+		goto err;
+	if (hexlenl(p, &extra) != 16 * 2 || extra)
+		goto err;
+
+	MEM_FREE(keeptr);
+	return 1;
+
+err:
+	MEM_FREE(keeptr);
+	return 0;
+}
+
+static void *get_salt(char *ciphertext)
+{
+	static struct custom_salt cs;
+	char *ctcopy = strdup(ciphertext);
+	char *keeptr = ctcopy;
+	char *p;
+	int i;
+
+	memset(&cs, 0, SALT_SIZE);
+	ctcopy += TAG_LENGTH;
+	p = strtokm(ctcopy, "*");
+	cs.type = atoi(p);
+	p = strtokm(NULL, "*");
+	for (i = 0; i < 16; i++)
+		cs.iv[i] = (atoi16[ARCH_INDEX(p[2 * i])] << 4) | atoi16[ARCH_INDEX(p[2 * i + 1])];
+	p = strtokm(NULL, "*");
+	for (i = 0; i < 16; i++)
+		cs.seed[i] = (atoi16[ARCH_INDEX(p[2 * i])] << 4) | atoi16[ARCH_INDEX(p[2 * i + 1])];
+
+	MEM_FREE(keeptr);
+	return &cs;
+}
+
+static void set_salt(void *salt)
+{
+	cur_salt = (struct custom_salt *)salt;
+}
+
+static void electrum_set_key(char *key, int index)
+{
+	strnzcpy(saved_key[index], key, PLAINTEXT_LENGTH + 1);
+}
+
+static char *get_key(int index)
+{
+	return saved_key[index];
+}
+
+static int crypt_all(int *pcount, struct db_salt *salt)
+{
+	const int count = *pcount;
+	int index = 0;
+
+	memset(cracked, 0, sizeof(cracked[0])*cracked_count);
+
+#ifdef _OPENMP
+#pragma omp parallel for
+	for (index = 0; index < count; index++)
+#endif
+	{
+		unsigned char iv[16];
+		unsigned char key[32];
+		unsigned char outbuf[48];
+		SHA256_CTX ctx;
+		AES_KEY aes_decrypt_key;
+		int extra;
+		int i;
+
+		SHA256_Init(&ctx);
+		SHA256_Update(&ctx, saved_key[index], strlen(saved_key[index]));
+		SHA256_Final(key, &ctx);
+		SHA256_Init(&ctx);
+		SHA256_Update(&ctx, key, 32);
+		SHA256_Final(key, &ctx);
+		memcpy(iv, cur_salt->iv, 16);
+		AES_set_decrypt_key(key, 128 * 2, &aes_decrypt_key);
+		AES_cbc_encrypt(cur_salt->seed, outbuf, 16, &aes_decrypt_key, iv, AES_DECRYPT);
+
+		if (cur_salt->type == 1) {
+			// check if 16 bytes of the encrypted seed are all lower-case hex (btcrecover)
+			outbuf[16] = 0;
+			if (hexlenl((const char*)outbuf, &extra) != 8 * 2 || extra)
+				cracked[index] = 0;
+			else
+				cracked[index] = 1;
+		} else if (cur_salt->type == 2) {
+			// check if starting 4 bytes are "xprv"
+			if (strncmp((const char*)outbuf, "xprv", 4))
+				cracked[index] = 0;
+			else {
+				// check if remaining 12 bytes are in base58 set [1-9A-HJ-NP-Za-km-z]
+				for (i = 0; i < 12; i++) {
+					unsigned char c = outbuf[4 + i];
+					if ((c > 'z') || (c < '1') || ((c > '9') && (c < 'A')) || ((c > 'Z') && (c < 'a'))) {
+						cracked[index] = 0;
+						break;
+					}
+				}
+				if (i == 12)
```

Comment:
@magnumripper I don't like this construct. Do you have a better way of writing it?